treeherder/perf/autoclassify.py

```diff
@@ -0,0 +1,84 @@
+    for i in range(int(summary.result_set_id), int(summary.prev_result_set_id), -1):
+        try:
+            result_set = get_result_set(summary.repository.name, i)
+        except IndexError:
+            return False
+        if "merge" in result_set["comments"].lower():
+            return True
+        for revision in result_set["revisions"]:
+            if "merge" in revision["comments"].lower():
+                return True
+    return False
+
+
+def get_revision_from_hg(summary):
+    """Get complete revision list from hg.mozilla.org"""
+    revisions = []
+    for i in range(int(summary.result_set_id), int(summary.prev_result_set_id), -1):
+        result_changeset = get_result_set(summary.repository.name, i)["revision"]
+        r = json.loads(make_request("https://hg.mozilla.org/integration/{}/json-pushes/?full=1&version=2&changeset={}".format(summary.repository.name, result_changeset)).text)
+        for push in r["pushes"].keys():
+            revisions.extend([changeset["node"] for changeset in r["pushes"][push]["changesets"]])
+    return revisions
+
+
+def classify_downstream(downstream, upstream_ids):
+    possible_upstreams = {}
+    revisions = get_revision_from_hg(downstream)
+    for alert in downstream.alerts.all():
+        if alert.id not in possible_upstreams.keys():
+            possible_upstreams[alert.id] = []
+        for upstream in PerformanceAlertSummary.objects.filter(
+                alerts__series_signature__signature_hash=alert.series_signature.signature_hash,
+                id__in=upstream_ids,
+                last_updated__gte=downstream.last_updated - timedelta(days=7)
+                ):
+            for i in range(int(upstream.result_set_id),
+                           int(upstream.prev_result_set_id),
+                           -1):
+                try:
+                    upstream_revisions = get_result_set(upstream.repository.name, i)
+                except IndexError:
+                    continue
+                for revision in upstream_revisions["revisions"]:
+                    if revision["revision"] in revisions:
+                        possible_upstreams[alert.id].append(upstream.id)
+        if len(set(possible_upstreams[alert.id])) == 1:
+            alert.status = 1
+            alert.related_summary_id = possible_upstreams[alert.id][0]
+            alert.save()
+    return possible_upstreams
+
+
+def classify():
+    alertsummaries = list(PerformanceAlertSummary.objects.order_by("-last_updated")[:100])
+
+    downstreams = [summary for summary in alertsummaries if is_downstream(summary)]
+
+    upstream_ids = [summary.id for summary in alertsummaries if summary.id not in [downstream.id for downstream in downstreams]]
+
+    mapping = {}
+    for downstream in downstreams:
+        mapping[downstream.id] = classify_downstream(downstream, upstream_ids)
+    for downstream, value in mapping.iteritems():
+        print "Downstream: {}".format(downstream)
```

Comment:
This is Python 3 incompatible, using `print()` is best practice when writing Python 2.7 code now :-)